hack/generator/pkg/astmodel/struct_definition.go

```diff
@@ -165,13 +159,26 @@ func (definition *StructDefinition) AsDeclarations() []ast.Decl {
 		declarations = append(declarations, resourceDeclaration)
 	}
 
+	// Append the methods
+	declarations = append(declarations, definition.generateMethodDecls()...)
+
 	return declarations
 }
 
+func (definition *StructDefinition) generateMethodDecls() []ast.Decl {
+	var result []ast.Decl
+	for methodName, function := range definition.StructType.functions {
+		funcDef := function.AsFunc(&definition.StructReference, methodName)
+		result = append(result, funcDef)
+	}
+
+	return result
+}
+
 // Tidy the content of this struct before generating the AST
 func (definition *StructDefinition) Tidy() {
-	sort.Slice(definition.fields, func(left int, right int) bool {
-		return definition.fields[left].fieldName < definition.fields[right].fieldName
+	sort.Slice(definition.StructType.fields, func(left int, right int) bool {
+		return definition.StructType.fields[left].fieldName < definition.StructType.fields[right].fieldName
 	})
```

Comment:
Given that `StructType` now incluces more than just fields, probably best if this method just delegates to a `Tidy()` method on the `StructType` itself - the methods will need to be sorted too, else we'll end up with diffs caused by them coming out in different orders.

Reply:
Our interface definitions have `Tidy` on `Definition` though, and right now there are no `Type`'s which have a `Tidy`, so I am not sure we want to violate that pattern?

Reply:
We have `Tidy()` on `Definition` to force it, not to prevent it elsewhere. Maybe it needs a different name?

The principle in the back of my mind is that types should be responsible for their own state; having `StructDefinition` reach in to monkey around with the internals of `StructType` isn't desirable.

Reply:
I don't really disagree, just pointing out that if you look at the impl of `Tidy`, 2/3 of them reach into the underlying `Type` and fuss with it.

Specifically, `EnumDefinition` and `StructDefinition` both do - I'm just wondering if the solution to the problem is a separate PR addressing `Tidy` specifically -- although here I probably should at least impl it for functions in `StructType` and I'm not sure how to as I don't know if go maps are ordered...

Reply:
Also as an aside, isn't `Tidy` in `StructDefinition` today violating `StructTypes` "immutability" guarantees (since it's sorting the contents in-place?)

It is _really_ awkward that there's not at least a convention to indicate "Don't fuss with my fields" -- I guess the solution is to put it in a different package...

Reply:
That's why I put the first model files into `astmodel` instead of next to the existing code.

Reply:
I have sent PR: #106 which removes Tidy entirely and just relies on immutability and does the sorting on create.

Reply:
If folks approve that I will rebase onto it and then add functions to it

Reply:
Did this